core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java

```diff
@@ -135,18 +137,29 @@ else if (partitioning.equals(FIXED_PASSTHROUGH_DISTRIBUTION)) {
             };
         }
         else if (partitioning.equals(FIXED_HASH_DISTRIBUTION) || partitioning.getConnectorId().isPresent()) {
-            exchangerSupplier = () -> new PartitioningExchanger(
-                    buffers,
-                    memoryManager,
-                    createPartitionFunction(
-                            nodePartitioningManager,
-                            session,
-                            blockTypeOperators,
-                            partitioning,
-                            bufferCount,
-                            partitionChannels,
-                            partitionChannelTypes,
-                            partitionHashChannel));
+            exchangerSupplier = () -> {
+                PartitionFunction partitionFunction = createPartitionFunction(
+                        nodePartitioningManager,
+                        session,
+                        blockTypeOperators,
+                        partitioning,
+                        bufferCount,
+                        partitionChannels,
+                        partitionChannelTypes,
+                        partitionHashChannel);
+                Function<Page, Page> partitionPagePreparer;
+                if (isSystemPartitioning(partitioning)) {
+                    partitionPagePreparer = identity();
+                }
+                else {
+                    partitionPagePreparer = page -> page.getColumns(Ints.toArray(partitionChannels));
```

Comment:
toArray can be lifted into its own line to avoid being called once per Page. Alternatively, couldn’t the partition channels and types be translated so that pages could be passed into the delegate directly?

Reply:
> Alternatively, couldn’t the partition channels and types be translated so that pages could be passed into the delegate directly?

I didn't undestand.

> toArray can be lifted into its own line to avoid being called once per Page.

that makes sense

Reply:
> Alternatively, couldn’t the partition channels and types be translated so that pages could be passed into the delegate directly?

That's possible. We could pass cannels into `io.trino.spi.connector.ConnectorNodePartitioningProvider#getBucketFunction` so that `BucketFunction` doesn't need to consume Page starting from channel 0. 

Since that would be a SPI change, we should understand the benefit we could get out of it.

Reply:
Got it, that makes sense. At the time I wasn’t aware that there was an assumption around 0-based indexing on the delegate and certainly didn’t appreciate the fact that it required passing the channel mask over the SPI boundary.